### PR TITLE
feat: add new filters to go bucketing logic

### PR DESCRIPTION
**Pull request by suthar26**
- adds new filters `startWith` and `endWith` for string comparison

### CODE REVIEW

File: bucketing/constants.go [hunk 1]
```diff
@@ -46,6 +46,10 @@ const (
 	ComparatorNotExist     = "!exist"
 	ComparatorContain      = "contain"
 	ComparatorNotContain   = "!contain"
+	ComparatorStartWith    = "startWith"
```
Comment (JamieSinn) [inline bucketing/constants.go:49]: is it `startWith` or `startsWith`

File: bucketing/segmentation.go [hunk 2]
```diff
@@ -103,6 +103,14 @@ func checkStringsFilter(str string, filter *UserFilter) bool {
 		return str != "" && stringArrayContains(values, str)
 	} else if operator == "!contain" {
 		return str == "" || !stringArrayContains(values, str)
+	} else if operator == "startWith" {
```
Comment (JamieSinn) [inline bucketing/segmentation.go:106]: Use the constant refs instead of the string raw itself.

File: bucketing/segmentation.go [hunk 3]
```diff
@@ -126,6 +134,24 @@ func stringArrayContains(substrings []string, search string) bool {
 	return false
 }
 
+func stringArrayStartsWith(prefixes []string, search string) bool {
+	for _, prefix := range prefixes {
+		if strings.HasPrefix(search, prefix) {
```
Comment (JamieSinn) [inline bucketing/segmentation.go:139]: Keep in mind - that `HasPrefix` and `HasSuffix` will return true if `prefix` in this arg is emptystring. 

This is something we should ensure is documented, or prevented from happening


`fmt.Println(strings.HasPrefix("Gopher", ""))`

Comment (suthar26) [reply bucketing/segmentation.go:139]: we dont currently catch this for `strings.Contains` either so if we add a check for `HasSuffix` and `HasPrefix` it should be added to `Contains` as well.. do you think that is a breaking change?

File: bucketing/constants.go (hunk 1 above)
Comment (suthar26) [reply bucketing/constants.go:49]: its `startWith` and `endWith`

File: bucketing/segmentation.go (hunk 3 above)
Comment (JamieSinn) [reply bucketing/segmentation.go:139]: as long as the bucketing logic is identical across platforms - we're fine - how does JS/WASM handle it?

File: bucketing/segmentation.go [hunk 4]
```diff
@@ -119,7 +127,25 @@ func stringArrayIn(arr []string, search string) bool {
 
 func stringArrayContains(substrings []string, search string) bool {
 	for _, substring := range substrings {
-		if strings.Contains(search, substring) {
+		if substring != "" && strings.Contains(search, substring) {
```
Comment (JamieSinn) [inline bucketing/segmentation.go:130]: You can just make this a continue if it's empty string, easier and cleaner

**Review by JamieSinn — APPROVED** [review]

**Review by JamieSinn — APPROVED** [review]

**Pull request merged**